source/mir/math/stat.d

```diff
@@ -23,99 +23,160 @@ import mir.math.common: fmamath;
 import mir.math.sum;
 import mir.primitives;
 import std.range.primitives: isInputRange;
-import std.traits: isArray, isFloatingPoint;
+import std.traits: isArray, isFloatingPoint, isMutable;
 
-private U prod(T : U[N], U, size_t N)(T input)
-    if (N > 0)
+/++
+Output range for mean.
++/
+struct MeanAccumulator(T, Summation summation) 
+    if (isMutable!T)
 {
-    static if (N == 1) {
-        return input[0];
-    } else {
-        U output = input[0];
-        for (size_t i = 1; i < N; i++)
-            output *= input[i];
-        return output;
+    private:
+    
+    Summator!(T, summation) sum;
+    size_t count;
+    
+    public:
+    
+    this()(T x, size_t n = 0) {
+        sum = Summator!(T, summation)(x);
+        count = n;
+    }
+    
+    F mean(F = T)() @property
+    {
+        return sum.sum / cast(F) count;
+    }
+    
+    void put(U)(U x)
+    {
+        this.sum.put(x);
+		size_t n;
+        static if (hasShape!U) {
+        	n = x.elementCount;
+        } else static if (hasLength!U) {
```

Comment:
If something has length, it is always has a shape property (if mir.primitives is included)